### PR TITLE
Add an ISO8601 serializer for Duration

### PR DESCRIPTION
**Pull request by jadengis**
I was using built_value to interface with a Spring server serving Durations in the ISO8601 format as described [here](https://en.wikipedia.org/wiki/ISO_8601#Durations). The microsecond Duration serializer was insufficient for handling this data.

I wrote a serializer in a in a similar vein to the already existing `Iso8601DateTimeSerializer` to handle these formatted durations.

This serializer does not support the full range of acceptable ISO8601 durations due to the limitations of the Dart `Duration` object (i.e. no year or week settings), but works for the range of values covered by
the `Duration` object.

This serializer is complete with tests similar to the ones for `Iso8601DateTimeSerializer`, and cover a lot of the edge cases, but I would be happy to add more if need be.

Please let me know if there is anything else you would like me to change.

### CODE REVIEW

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+/// Note that this serializer is not 100% compatible with the ISO8601 format
+/// due to limitations of the [Duration] class, but is designed to produce and
+/// consume reasonable strings that match the standard.
+class Iso8601DurationSerializer extends PrimitiveSerializer<Duration> {
+  @override
+  Duration deserialize(Serializers serializers, Object serialized,
+      {FullType specifiedType = FullType.unspecified}) {
+    final String value = serialized;
+    return _parseDuration(value);
+  }
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: You don't need `value`; this can just be a one liner method with =>.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  @override
+  Object serialize(Serializers serializers, Duration object,
+      {FullType specifiedType = FullType.unspecified}) {
+    return _writeIso8601Duration(object);
+  }
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: Please use =>.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        unitMappings[unit] = value;
+      }
+      return Duration(
+        days: unitMappings[_dayToken] ?? 0,
+        hours: unitMappings[_hourToken] ?? 0,
+        minutes: unitMappings[_minuteToken] ?? 0,
+        seconds: unitMappings[_secondToken] ?? 0,
+      );
+    } else {
+      throw FormatException("Invalid duration format", value);
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: Please move this up as a 'fail early' to reduce indentation.

if (match == null) throw FormatException(...).

Additionally, please use single quotes for strings.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  @override
+  Iterable<Type> get types => BuiltList(const [Duration]);
+
+  @override
+  String get wireName => 'Duration';
+
+  Duration _parseDuration(String value) {
+    final match = _parseFormat.firstMatch(value);
+    if (match != null) {
+      // Iterate through the capture groups to build the unit mappings
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: https://www.dartlang.org/guides/language/effective-dart/documentation#do-format-comments-like-sentences

here and below.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      final unitMappings = <String, int>{};
+
+      // Start iterating at 1, because match[0] is the full match
+      for (var i = 1; i <= match.groupCount; i++) {
+        final group = match[i];
+        if (group == null) {
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: This can be a single line if you like. Up to you.

if (group == null) continue;

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        seconds: unitMappings[_secondToken] ?? 0,
+      );
+    } else {
+      throw FormatException("Invalid duration format", value);
+    }
+  }
+
+  String _writeIso8601Duration(Duration duration) {
+    if (duration == Duration.zero) {
+      // format for zero.
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: I don't think this comment adds anything. Up to you.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            .inSeconds;
+    final buffer = StringBuffer(_durationToken)
+      ..write(days == 0 ? '' : '$days$_dayToken');
+    if (!(hours == 0 && minutes == 0 && seconds == 0)) {
+      buffer
+        ..write(_timeToken)
+        ..write(hours == 0 ? '' : '$hours$_hourToken')
+        ..write(minutes == 0 ? '' : '$minutes$_minuteToken')
+        ..write(seconds == 0 ? '' : '$seconds$_secondToken');
+    }
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: Does the format support subsecond precision?

If not, it would probably be best to throw an exception if there is anything left over that can't be serialized, i.e. any sub-second data.

File: built_value/lib/iso_8601_duration_serializer.dart [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  static const _durationToken = 'P';
+  static const _dayToken = 'D';
+  static const _timeToken = 'T';
+  static const _hourToken = 'H';
+  static const _minuteToken = 'M';
+  static const _secondToken = 'S';
+
+  // Parse format for ISO8601
+  static final _parseFormat = RegExp(
+      '^P(?!\$)(0D|[1-9][0-9]*D)?(?:T(?!\$)(0H|[1-9][0-9]*H)?(0M|[1-9][0-9]*M)?(0S|[1-9][0-9]*S)?)?\$');
```
Comment (davidmorgan) [inline built_value/lib/iso_8601_duration_serializer.dart]: Please split the string so this fits in 80 cols.

File: built_value/lib/iso_8601_duration_serializer.dart (hunk 7 above)
Comment (jadengis) [reply built_value/lib/iso_8601_duration_serializer.dart]: The format does not support the sub-second precision. I'll add the safety check.

**Comment by jadengis** [comment]
@davidmorgan I addressed you comments an updated the CHANGELOG. Let me know if the new version number is okay.

**Comment by davidmorgan** [comment]
Looks good, thanks!
